docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+}
+```
+
+## Use high-level cross-contract API
+
+There is a helper macro that allows you to make cross-contract calls with the syntax `#[ext_contract(...)]`. It takes a Rust Trait and
+converts it to a module with static methods. Each of these static methods takes positional arguments defined by the Trait,
+then the `receiver_id`, the attached deposit and the amount of gas and returns a new `Promise`.
+
+For example, let's define a calculator contract Trait:
+
+```rust
+#[ext_contract(ext_calculator)]
+trait Calculator {
+    fn mult(&self, a: U64, b: U64) -> U128;
+
+    fn sum(&self, a: U128, b: U128) -> U128;
+}
+```
+
+It's equivalent to the following code:
+
+```rust
+mod ext_calculator {
+    pub fn mult(a: U64, b: U64, receiver_id: &AccountId, deposit: Balance, gas: Gas) -> Promise {
+        Promise::new(receiver_id.clone())
+            .function_call(
+                b"mult",
+                json!({ "a": a, "b": b }).to_string().as_bytes(),
+                deposit,
+                gas,
+            )
+    }
+
+    pub fn sum(a: U128, b: U128, receiver_id: &AccountId, deposit: Balance, gas: Gas) -> Promise {
+        // ...
+    }
+}
+```
+
+Let's assume the calculator is deployed on `calc.near`, we can use the following:
+
+```rust
+const CALCULATOR_ACCOUNT_ID: &str = "calc.near";
+const NO_DEPOSIT: Balance = 0;
+const BASE_GAS: Gas = 5_000_000_000_000;
+
+#[near_bindgen]
+impl Contract {
+    pub fn sum_a_b(&mut self, a: U128, b: U128) -> Promise {
+        let calculator_account_id: AccountId = CALCULATOR_ACCOUNT_ID.to_string();
+        ext_calculator::sum(a, b, &calculator_account_id, NO_DEPOSIT, BASE_GAS)
+    }
+}
+```
+
+## Reuse crates from `near-sdk`
+
+`near-sdk` re-exports the following crates:
+
+- `borsh`
+- `base64`
+- `bs58`
+- `serde`
+- `serde_json`
+
+Most common crates include `borsh` which is needed for internal STATE serialization and
+`serde` for external JSON serialization.
+
+When marking structs with `serde::Serialize` you need to use `#[serde(crate = "near_sdk::serde")]`
+to point serde to the correct base crate.
+
+```rust
+/// Import `borsh` from `near_sdk` crate 
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+/// Import `serde` from `near_sdk` crate 
+use near_sdk::serde::{Serialize, Deserialize};
+
+/// Main contract structure serialized with Borsh
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub pair: Pair,
+}
+
+/// Implements both `serde` and `borsh` serialization.
+/// `serde` is typically useful when returning a struct in JSON format for a frontend.
+#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Pair {
+    pub a: u32,
+    pub b: u32,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(pair: Pair) -> Self {
+        Self {
+            pair,
+        }
+    }
+
+    pub fn get_pair(self) -> Pair {
+        self.pair
+    }
+}
+```
```

Comment:
imo we shouldn't document this, I don't think it's a pattern we should enforce or keep in the future.

Reply:
feels like we may need to document the `#[serde(crate = "near_sdk::serde")]` part in particular. I have found myself frequently looking for this line as I can't seem to remember it. I am tempted to keep this so it's discoverable, and we can remove it when the SDK changes? what do you think?

Reply:
Yeah maybe that's a good idea, my opinion is that the pattern of re-exporting from `near_sdk` at all, especially for serde which is unlikely to see a new major version, is a bit weird and I'd rather move the direction of using serde as a dependency and not requiring this.

As for the others, I really don't see the motivation in the re-exports, is it to avoid users from using the crates in their code and potentially having duplicate dependencies for these if the versions in the SDK are updated?

I think fine to keep docs, at least for now, was just sharing an opinion

Reply:
Interesting, and thanks for the context. I am not sure why we included these either, to be honest. I'd say let's move towards getting this more "normal" but provide this section of the SDK Docs for folks that want to copy/paste some logic contained in here. I know I certainly have found it useful.